applications/MedApplication/custom_utilities/med_testing_utilities.cpp

```diff
@@ -0,0 +1,197 @@
+using GeometryType = ModelPart::GeometryType;
+using GeometryContainerType = ModelPart::GeometryContainerType;
+
+template<class T>
+bool contains(
+    const std::vector<T> rVec,
+    const T& rValue)
+{
+    return std::find(rVec.begin(), rVec.end(), rValue) != rVec.end();
+}
+
+// check if the geometries have the same nodes
+// for this to be the case, two conditions need to be fulflled:
+// 1: geometries have same number of nodes
+// 2: nodes of both geometries are in same order and in same coords
+bool have_same_nodes(
+    const GeometryType& rGeom1,
+    const GeometryType& rGeom2)
+{
+    if (rGeom1.PointsNumber() != rGeom2.PointsNumber()) return false;
+
+    for (std::size_t i=0; i<rGeom1.PointsNumber(); ++i) {
+        if (rGeom1[i].Distance(rGeom2[i]) > 1e-12) return false;
+    }
+
+    return true;
+}
+
+void CheckEntitiesAreEqual(
+    const NodeType& rNode1,
+    const NodeType& rNode2)
+{
+    KRATOS_TRY
+
+    // KRATOS_CHECK_EQUAL(rNode1.Id(), rNode2.Id()); // the MEDApp deliberately does not care about IDs
+
+    KRATOS_CHECK_DOUBLE_EQUAL(rNode1.X(),  rNode2.X());
+    KRATOS_CHECK_DOUBLE_EQUAL(rNode1.X0(), rNode2.X0());
+
+    KRATOS_CHECK_DOUBLE_EQUAL(rNode1.Y(),  rNode2.Y());
+    KRATOS_CHECK_DOUBLE_EQUAL(rNode1.Y0(), rNode2.Y0());
+
+    KRATOS_CHECK_DOUBLE_EQUAL(rNode1.Z(),  rNode2.Z());
+    KRATOS_CHECK_DOUBLE_EQUAL(rNode1.Z0(), rNode2.Z0());
+
+    KRATOS_CATCH("")
+}
+
+void CheckGeometriesAreEqual(
+    const GeometryType& rGeom1,
+    const GeometryType& rGeom2)
+{
+    KRATOS_TRY
+
+    // KRATOS_CHECK_EQUAL(rGeom1.Id(), rGeom2.Id()); // the MEDApp deliberately does not care about IDs
+
+    KRATOS_CHECK_EQUAL(rGeom1.PointsNumber(), rGeom2.PointsNumber());
+
+    KRATOS_CHECK(GeometryType::IsSame(rGeom1, rGeom2));
+
+    for (std::size_t i=0; i<rGeom1.PointsNumber(); ++i) {
+        CheckEntitiesAreEqual(rGeom1[i], rGeom2[i]);
+    }
+
+    KRATOS_CATCH("")
+}
+
+template<class TContainerType>
+void CheckEntitiesAreEqual(
+    const TContainerType& rEntities1,
+    const TContainerType& rEntities2)
+{
+    KRATOS_TRY
+
+    // basic checks
+    KRATOS_CHECK_EQUAL(rEntities1.size(), rEntities2.size());
+
+    // check entities
+    for (std::size_t i=0; i<rEntities1.size(); ++i) {
+        CheckEntitiesAreEqual(*(rEntities1.begin()+i), *(rEntities2.begin()+i));
+    }
+
+    KRATOS_CATCH("")
+}
+
+void CheckGeometriesAreEqual(
+    const ModelPart& rModelPart1,
+    const ModelPart& rModelPart2)
+{
+    KRATOS_TRY
+
+    // basic checks
+    KRATOS_CHECK_EQUAL(rModelPart1.NumberOfGeometries(), rModelPart2.NumberOfGeometries());
+
+    auto check_geoms = [](const ModelPart& rModelPart1, const ModelPart& rModelPart2){
+        for (auto geom_1 : rModelPart1.Geometries()) {
+            for (auto geom_2 : rModelPart2.Geometries()) {
+                if (have_same_nodes(geom_1, geom_2)) {
+                    CheckGeometriesAreEqual(geom_1, geom_2);
+                    goto here;
```

Comment:
`break`?

Reply:
I am pretty sure I tried, but `break` IIRC works only for the inner loop

Reply:
https://stackoverflow.com/questions/1257744/can-i-use-break-to-exit-multiple-nested-for-loops

Reply:
Unless I'm missing something, you're just breaking from the inner loop, not the outer one.

Reply:
Seems so, but I also know that implementing this took me a while (first time I eve used `goto`)

Since I am a bit out of the loop rn I will leave it as is and revisit it next time I develop